{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "meth24_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode24.CG.bed', sep='\\t', header=None)\n",
    "meth24_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode24.CHG.bed', sep='\\t', header=None)\n",
    "meth24_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode24.CHH.bed', sep='\\t', header=None)\n",
    "meth01_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode01.CG.bed', sep='\\t', header=None)\n",
    "meth01_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode01.CHG.bed', sep='\\t', header=None)\n",
    "meth01_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode01.CHH.bed', sep='\\t', header=None)\n",
    "meth25_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode25.CG.bed', sep='\\t', header=None)\n",
    "meth25_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode25.CHG.bed', sep='\\t', header=None)\n",
    "meth25_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode25.CHH.bed', sep='\\t', header=None)\n",
    "meth29_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode29.CG.bed', sep='\\t', header=None)\n",
    "meth29_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode29.CHG.bed', sep='\\t', header=None)\n",
    "meth29_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode29.CHH.bed', sep='\\t', header=None)\n",
    "meth19_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode19.CG.bed', sep='\\t', header=None)\n",
    "meth19_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode19.CHG.bed', sep='\\t', header=None)\n",
    "meth19_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode19.CHH.bed', sep='\\t', header=None)\n",
    "meth26_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode26.CG.bed', sep='\\t', header=None)\n",
    "meth26_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode26.CHG.bed', sep='\\t', header=None)\n",
    "meth26_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode26.CHH.bed', sep='\\t', header=None)\n",
    "meth20_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode20.CG.bed', sep='\\t', header=None)\n",
    "meth20_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode20.CHG.bed', sep='\\t', header=None)\n",
    "meth20_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode20.CHH.bed', sep='\\t', header=None)\n",
    "meth27_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode27.CG.bed', sep='\\t', header=None)\n",
    "meth27_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode27.CHG.bed', sep='\\t', header=None)\n",
    "meth27_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode27.CHH.bed', sep='\\t', header=None)\n",
    "meth21_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode21.CG.bed', sep='\\t', header=None)\n",
    "meth21_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode21.CHG.bed', sep='\\t', header=None)\n",
    "meth21_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode21.CHH.bed', sep='\\t', header=None)\n",
    "meth28_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode28.CG.bed', sep='\\t', header=None)\n",
    "meth28_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode28.CHG.bed', sep='\\t', header=None)\n",
    "meth28_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode28.CHH.bed', sep='\\t', header=None)\n",
    "meth22_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode22.CG.bed', sep='\\t', header=None)\n",
    "meth22_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode22.CHG.bed', sep='\\t', header=None)\n",
    "meth22_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode22.CHH.bed', sep='\\t', header=None)\n",
    "meth30_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode30.CG.bed', sep='\\t', header=None)\n",
    "meth30_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode30.CHG.bed', sep='\\t', header=None)\n",
    "meth30_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode30.CHH.bed', sep='\\t', header=None)\n",
    "meth23_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode23.CG.bed', sep='\\t', header=None)\n",
    "meth23_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode23.CHG.bed', sep='\\t', header=None)\n",
    "meth23_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode23.CHH.bed', sep='\\t', header=None)\n",
    "meth31_cg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode31.CG.bed', sep='\\t', header=None)\n",
    "meth31_chg = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode31.CHG.bed', sep='\\t', header=None)\n",
    "meth31_chh = pd.read_csv('../ont_methylation/SQK-NBD114-96_barcode31.CHH.bed', sep='\\t', header=None)\n",
    "\n",
    "df_list = [meth01_cg, meth01_chg, meth01_chh, \n",
    "           meth29_cg, meth29_chg, meth29_chh, \n",
    "           meth24_cg, meth24_chg, meth24_chh, \n",
    "           meth25_cg, meth25_chg, meth25_chh,\n",
    "           meth19_cg, meth19_chg, meth19_chh, \n",
    "           meth26_cg, meth26_chg, meth26_chh, \n",
    "           meth20_cg, meth20_chg, meth20_chh, \n",
    "           meth27_cg, meth27_chg, meth27_chh, \n",
    "           meth21_cg, meth21_chg, meth21_chh, \n",
    "           meth28_cg, meth28_chg, meth28_chh, \n",
    "           meth22_cg, meth22_chg, meth22_chh, \n",
    "           meth30_cg, meth30_chg, meth30_chh, \n",
    "           meth23_cg, meth23_chg, meth23_chh, \n",
    "           meth31_cg, meth31_chg, meth31_chh]\n",
    "\n",
    "name_list = ['meth01_cg', 'meth01_chg', 'meth01_chh', \n",
    "           'meth29_cg', 'meth29_chg', 'meth29_chh', \n",
    "           'meth24_cg', 'meth24_chg', 'meth24_chh', \n",
    "           'meth25_cg', 'meth25_chg', 'meth25_chh', \n",
    "           'meth19_cg', 'meth19_chg', 'meth19_chh', \n",
    "           'meth26_cg', 'meth26_chg', 'meth26_chh', \n",
    "           'meth20_cg', 'meth20_chg', 'meth20_chh', \n",
    "           'meth27_cg', 'meth27_chg', 'meth27_chh', \n",
    "           'meth21_cg', 'meth21_chg', 'meth21_chh', \n",
    "           'meth28_cg', 'meth28_chg', 'meth28_chh', \n",
    "           'meth22_cg', 'meth22_chg', 'meth22_chh', \n",
    "           'meth30_cg', 'meth30_chg', 'meth30_chh', \n",
    "           'meth23_cg', 'meth23_chg', 'meth23_chh', \n",
    "           'meth31_cg', 'meth31_chg', 'meth31_chh']\n",
    "filt_df = {}\n",
    "\n",
    "for df, name in zip(df_list, name_list):\n",
    "    filt_df[name] = df[(df[9] <= df[9].quantile(0.95)) & (df[9] >= df[9].quantile(0.05))]\n",
    "\n",
    "meth24_cg = filt_df['meth24_cg']\n",
    "meth24_chg = filt_df['meth24_chg']\n",
    "meth24_chh = filt_df['meth24_chh']\n",
    "meth01_cg = filt_df['meth01_cg']\n",
    "meth01_chg = filt_df['meth01_chg']\n",
    "meth01_chh = filt_df['meth01_chh']\n",
    "meth25_cg = filt_df['meth25_cg']\n",
    "meth25_chg = filt_df['meth25_chg']\n",
    "meth25_chh = filt_df['meth25_chh']\n",
    "meth29_cg = filt_df['meth29_cg']\n",
    "meth29_chg = filt_df['meth29_chg']\n",
    "meth29_chh = filt_df['meth29_chh']\n",
    "meth19_cg = filt_df['meth19_cg']\n",
    "meth19_chg = filt_df['meth19_chg']\n",
    "meth19_chh = filt_df['meth19_chh']\n",
    "meth26_cg = filt_df['meth26_cg']\n",
    "meth26_chg = filt_df['meth26_chg']\n",
    "meth26_chh = filt_df['meth26_chh']\n",
    "meth20_cg = filt_df['meth20_cg']\n",
    "meth20_chg = filt_df['meth20_chg']\n",
    "meth20_chh = filt_df['meth20_chh']\n",
    "meth27_cg = filt_df['meth27_cg']\n",
    "meth27_chg = filt_df['meth27_chg']\n",
    "meth27_chh = filt_df['meth27_chh']\n",
    "meth21_cg = filt_df['meth21_cg']\n",
    "meth21_chg = filt_df['meth21_chg']\n",
    "meth21_chh = filt_df['meth21_chh']\n",
    "meth28_cg = filt_df['meth28_cg']\n",
    "meth28_chg = filt_df['meth28_chg']\n",
    "meth28_chh = filt_df['meth28_chh']\n",
    "meth22_cg = filt_df['meth22_cg']\n",
    "meth22_chg = filt_df['meth22_chg']\n",
    "meth22_chh = filt_df['meth22_chh']\n",
    "meth30_cg = filt_df['meth30_cg']\n",
    "meth30_chg = filt_df['meth30_chg']\n",
    "meth30_chh = filt_df['meth30_chh']\n",
    "meth23_cg = filt_df['meth23_cg']\n",
    "meth23_chg = filt_df['meth23_chg']\n",
    "meth23_chh = filt_df['meth23_chh']\n",
    "meth31_cg = filt_df['meth31_cg']\n",
    "meth31_chg = filt_df['meth31_chg']\n",
    "meth31_chh = filt_df['meth31_chh']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "onsen24 = pd.read_csv('onsen_positions/ONSEN_barcode24.gff', sep='\\t', header=None)\n",
    "onsen01 = pd.read_csv('onsen_positions/ONSEN_barcode01.gff', sep='\\t', header=None)\n",
    "onsen25 = pd.read_csv('onsen_positions/ONSEN_barcode25.gff', sep='\\t', header=None)\n",
    "onsen29 = pd.read_csv('onsen_positions/ONSEN_barcode29.gff', sep='\\t', header=None)\n",
    "onsen19 = pd.read_csv('onsen_positions/ONSEN_barcode19.gff', sep='\\t', header=None)\n",
    "onsen26 = pd.read_csv('onsen_positions/ONSEN_barcode26.gff', sep='\\t', header=None)\n",
    "onsen20 = pd.read_csv('onsen_positions/ONSEN_barcode20.gff', sep='\\t', header=None)\n",
    "onsen27 = pd.read_csv('onsen_positions/ONSEN_barcode27.gff', sep='\\t', header=None)\n",
    "onsen21 = pd.read_csv('onsen_positions/ONSEN_barcode21.gff', sep='\\t', header=None)\n",
    "onsen28 = pd.read_csv('onsen_positions/ONSEN_barcode28.gff', sep='\\t', header=None)\n",
    "onsen22 = pd.read_csv('onsen_positions/ONSEN_barcode22.gff', sep='\\t', header=None)\n",
    "onsen30 = pd.read_csv('onsen_positions/ONSEN_barcode30.gff', sep='\\t', header=None)\n",
    "onsen23 = pd.read_csv('onsen_positions/ONSEN_barcode23.gff', sep='\\t', header=None)\n",
    "onsen31 = pd.read_csv('onsen_positions/ONSEN_barcode31.gff', sep='\\t', header=None)\n",
    "\n",
    "# get only full-length ONSENs\n",
    "onsen24 = onsen24[onsen24.loc[:, 4] - onsen24.loc[:, 3] >= 3000] \n",
    "onsen01 = onsen01[onsen01.loc[:, 4] - onsen01.loc[:, 3] >= 3000] \n",
    "onsen25 = onsen25[onsen25.loc[:, 4] - onsen25.loc[:, 3] >= 3000] \n",
    "onsen29 = onsen29[onsen29.loc[:, 4] - onsen29.loc[:, 3] >= 3000] \n",
    "onsen19 = onsen19[onsen19.loc[:, 4] - onsen19.loc[:, 3] >= 3000] \n",
    "onsen26 = onsen26[onsen26.loc[:, 4] - onsen26.loc[:, 3] >= 3000] \n",
    "onsen20 = onsen20[onsen20.loc[:, 4] - onsen20.loc[:, 3] >= 3000] \n",
    "onsen27 = onsen27[onsen27.loc[:, 4] - onsen27.loc[:, 3] >= 3000] \n",
    "onsen21 = onsen21[onsen21.loc[:, 4] - onsen21.loc[:, 3] >= 3000] \n",
    "onsen28 = onsen28[onsen28.loc[:, 4] - onsen28.loc[:, 3] >= 3000] \n",
    "onsen22 = onsen22[onsen22.loc[:, 4] - onsen22.loc[:, 3] >= 3000] \n",
    "onsen30 = onsen30[onsen30.loc[:, 4] - onsen30.loc[:, 3] >= 3000] \n",
    "onsen23 = onsen23[onsen23.loc[:, 4] - onsen23.loc[:, 3] >= 3000] \n",
    "onsen31 = onsen31[onsen31.loc[:, 4] - onsen31.loc[:, 3] >= 3000] \n",
    "\n",
    "# shift start of ONSEN7\n",
    "onsen24.loc[3, 3] += 3118 - 427\n",
    "onsen01.loc[3, 3] += 3118 - 427\n",
    "onsen25.loc[3, 3] += 3118 - 427\n",
    "onsen29.loc[3, 3] += 3118 - 427\n",
    "onsen19.loc[3, 3] += 3118 - 427\n",
    "onsen26.loc[3, 3] += 3118 - 427\n",
    "onsen20.loc[3, 3] += 3118 - 427\n",
    "onsen27.loc[3, 3] += 3118 - 427\n",
    "onsen21.loc[3, 3] += 3118 - 427\n",
    "onsen28.loc[3, 3] += 3118 - 427\n",
    "onsen22.loc[3, 3] += 3118 - 427\n",
    "onsen30.loc[3, 3] += 3118 - 427\n",
    "onsen23.loc[3, 3] += 3118 - 427\n",
    "onsen31.loc[3, 3] += 3118 - 427"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "nte01 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode01.bed', sep='\\t', header=None)\n",
    "nte29 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode29.bed', sep='\\t', header=None)\n",
    "nte19 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode19.bed', sep='\\t', header=None)\n",
    "nte26 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode26.bed', sep='\\t', header=None)\n",
    "nte20 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode20.bed', sep='\\t', header=None)\n",
    "nte27 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode27.bed', sep='\\t', header=None)\n",
    "nte21 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode21.bed', sep='\\t', header=None)\n",
    "nte28 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode28.bed', sep='\\t', header=None)\n",
    "nte22 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode22.bed', sep='\\t', header=None)\n",
    "nte30 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode30.bed', sep='\\t', header=None)\n",
    "nte23 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode23.bed', sep='\\t', header=None)\n",
    "nte31 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode31.bed', sep='\\t', header=None)\n",
    "nte24 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode24.bed', sep='\\t', header=None)\n",
    "nte25 = pd.read_csv('pseudoreference/novel_TEI_SQK-NBD114-96_barcode25.bed', sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dict = {\n",
    "    'Col-0_CG': (meth24_cg, onsen24, meth25_cg, onsen25),  # CG context, barcodes 24-25\n",
    "    'Col-0_CHG': (meth24_chg, onsen24, meth25_chg, onsen25),  # CHG context, barcodes 24-25\n",
    "    'Col-0_CHH': (meth24_chh, onsen24, meth25_chh, onsen25),   # CHH context, barcodes 24-25\n",
    "\n",
    "    '12-1_CG': (meth01_cg, onsen01, meth29_cg, onsen29),  # CG context, barcodes 01-29\n",
    "    '12-1_CHG': (meth01_chg, onsen01, meth29_chg, onsen29),  # CHG context, barcodes 01-29\n",
    "    '12-1_CHH': (meth01_chh, onsen01, meth29_chh, onsen29),  # CHH context, barcodes 01-29\n",
    "\n",
    "    '1-1_CG': (meth19_cg, onsen19, meth26_cg, onsen26), \n",
    "    '1-1_CHG': (meth19_chg, onsen19, meth26_chg, onsen26),  \n",
    "    '1-1_CHH': (meth19_chh, onsen19, meth26_chh, onsen26), \n",
    "\n",
    "    '3-6_CG': (meth20_cg, onsen20, meth27_cg, onsen27), \n",
    "    '3-6_CHG': (meth20_chg, onsen20, meth27_chg, onsen27),  \n",
    "    '3-6_CHH': (meth20_chh, onsen20, meth27_chh, onsen27), \n",
    "\n",
    "    '7-5_CG': (meth21_cg, onsen21, meth28_cg, onsen28), \n",
    "    '7-5_CHG': (meth21_chg, onsen21, meth28_chg, onsen28),  \n",
    "    '7-5_CHH': (meth21_chh, onsen21, meth28_chh, onsen28), \n",
    "\n",
    "    '13-1_CG': (meth22_cg, onsen22, meth30_cg, onsen30), \n",
    "    '13-1_CHG': (meth22_chg, onsen22, meth30_chg, onsen30),  \n",
    "    '13-1_CHH': (meth22_chh, onsen22, meth30_chh, onsen30), \n",
    "\n",
    "    '14-2_CG': (meth23_cg, onsen23, meth31_cg, onsen31), \n",
    "    '14-2_CHG': (meth23_chg, onsen23, meth31_chg, onsen31),  \n",
    "    '14-2_CHH': (meth23_chh, onsen23, meth31_chh, onsen31), \n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Binned methylation in original full-length ONSEN copies\n",
    "## TE body"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Col-0_CG\n",
      "Col-0_CHG\n",
      "Col-0_CHH\n",
      "12-1_CG\n",
      "12-1_CHG\n",
      "12-1_CHH\n",
      "1-1_CG\n",
      "1-1_CHG\n",
      "1-1_CHH\n",
      "3-6_CG\n",
      "3-6_CHG\n",
      "3-6_CHH\n",
      "7-5_CG\n",
      "7-5_CHG\n",
      "7-5_CHH\n",
      "13-1_CG\n",
      "13-1_CHG\n",
      "13-1_CHH\n",
      "14-2_CG\n",
      "14-2_CHG\n",
      "14-2_CHH\n"
     ]
    }
   ],
   "source": [
    "# Define the number of bins\n",
    "NUM_BINS = 100\n",
    "\n",
    "# Initialize a dictionary to store results for each sample and context\n",
    "results_dict = {}\n",
    "\n",
    "# Iterate through each sample and context\n",
    "for sample_context, (df1, onsen1, df2, onsen2) in data_dict.items():\n",
    "    # Initialize a list to store combined binned methylation for all ONSENs\n",
    "    combined_binned_methylation = []\n",
    "    \n",
    "    # Get unique ONSENs for each barcode\n",
    "    onsens_barcode1 = onsen1.drop_duplicates().reset_index(drop=True)\n",
    "    onsens_barcode2 = onsen2.drop_duplicates().reset_index(drop=True)\n",
    "    \n",
    "    # Iterate through ONSENs in barcode 1\n",
    "    for _, onsen_row in onsens_barcode1.iterrows():\n",
    "        # Extract ONSEN coordinates\n",
    "        chr_name = onsen_row[0]  # Chromosome\n",
    "        start = onsen_row[3]     # Start position\n",
    "        end = onsen_row[4]       # End position\n",
    "        onsen_name = onsen_row[8].split(';')[0].split('=')[1]  # Extract ONSEN name from column 8\n",
    "        \n",
    "        # Initialize lists to store combined methylation values and counts for each bin\n",
    "        combined_bins_meth = [[] for _ in range(NUM_BINS)]\n",
    "        combined_bins_count = [0] * NUM_BINS\n",
    "        \n",
    "        # Calculate bin size for the current ONSEN\n",
    "        bin_size = (end - start) / NUM_BINS\n",
    "        \n",
    "        # Process methylation data for barcode 1\n",
    "        onsen_meth_barcode1 = df1[\n",
    "            (df1[0] == chr_name) &  # Match chromosome\n",
    "            (df1[1] >= start) &     # Start position >= ONSEN start\n",
    "            (df1[2] <= end)         # End position <= ONSEN end\n",
    "        ]\n",
    "        for _, row in onsen_meth_barcode1.iterrows():\n",
    "            pos = row[1]  # Methylation position\n",
    "            bin_idx = min(int((pos - start) // bin_size), NUM_BINS - 1)  # Ensure bin index is within range\n",
    "            combined_bins_meth[bin_idx].append(row[10])  # Assuming column 10 contains methylation values\n",
    "            combined_bins_count[bin_idx] += 1  # Count cytosines\n",
    "        \n",
    "        # Process methylation data for barcode 2\n",
    "        onsen_meth_barcode2 = df2[\n",
    "            (df2[0] == chr_name) &  # Match chromosome\n",
    "            (df2[1] >= start) &     # Start position >= ONSEN start\n",
    "            (df2[2] <= end)         # End position <= ONSEN end\n",
    "        ]\n",
    "        for _, row in onsen_meth_barcode2.iterrows():\n",
    "            pos = row[1]  # Methylation position\n",
    "            bin_idx = min(int((pos - start) // bin_size), NUM_BINS - 1)  # Ensure bin index is within range\n",
    "            combined_bins_meth[bin_idx].append(row[10])  # Assuming column 10 contains methylation values\n",
    "            combined_bins_count[bin_idx] += 1  # Count cytosines\n",
    "        \n",
    "        # Calculate mean methylation for each bin\n",
    "        mean_binned_meth = []\n",
    "        for meth_values, count in zip(combined_bins_meth, combined_bins_count):\n",
    "            if count > 0:\n",
    "                mean_meth = np.mean(meth_values)  # Calculate mean methylation\n",
    "            else:\n",
    "                mean_meth = 0  # If no cytosines, set methylation to 0\n",
    "            mean_binned_meth.append(mean_meth)\n",
    "        \n",
    "        # Add ONSEN name and mean binned methylation to the results\n",
    "        combined_binned_methylation.append([onsen_name] + mean_binned_meth)\n",
    "    \n",
    "    # Convert the results to a DataFrame\n",
    "    columns = ['ONSEN'] + [f'bin_{i+1}' for i in range(NUM_BINS)]\n",
    "    results_df = pd.DataFrame(combined_binned_methylation, columns=columns)\n",
    "    \n",
    "    # Store the results in the dictionary\n",
    "    results_dict[sample_context] = results_df\n",
    "\n",
    "    # Save the results to CSV files\n",
    "    results_df.to_csv(f'ONSENs_meth_bins/ONSEN_binned_methylation_{sample_context}.csv', index=False)\n",
    "    print(sample_context)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TE flanks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Col-0_CG\n",
      "Col-0_CHG\n",
      "Col-0_CHH\n",
      "12-1_CG\n",
      "12-1_CHG\n",
      "12-1_CHH\n",
      "1-1_CG\n",
      "1-1_CHG\n",
      "1-1_CHH\n",
      "3-6_CG\n",
      "3-6_CHG\n",
      "3-6_CHH\n",
      "7-5_CG\n",
      "7-5_CHG\n",
      "7-5_CHH\n",
      "13-1_CG\n",
      "13-1_CHG\n",
      "13-1_CHH\n",
      "14-2_CG\n",
      "14-2_CHG\n",
      "14-2_CHH\n"
     ]
    }
   ],
   "source": [
    "# Define the bin size and flanking size\n",
    "BIN_SIZE = 50  # Fixed bin size\n",
    "FLANKING_SIZE = 2000  # 2000 bp flanking regions\n",
    "\n",
    "# Initialize a dictionary to store results for each sample and context\n",
    "results_flank_dict = {}\n",
    "\n",
    "# Iterate through each sample and context\n",
    "for sample_context, (df1, onsen1, df2, onsen2) in data_dict.items():\n",
    "    # Initialize a list to store combined binned methylation for all ONSENs\n",
    "    combined_binned_methylation = []\n",
    "    \n",
    "    # Get unique ONSENs for each barcode\n",
    "    onsens_barcode1 = onsen1.drop_duplicates().reset_index(drop=True)\n",
    "    onsens_barcode2 = onsen2.drop_duplicates().reset_index(drop=True)\n",
    "    \n",
    "    # Iterate through ONSENs in barcode 1\n",
    "    for _, onsen_row in onsens_barcode1.iterrows():\n",
    "        # Extract ONSEN coordinates\n",
    "        chr_name = onsen_row[0]  # Chromosome\n",
    "        start = onsen_row[3]     # Start position\n",
    "        end = onsen_row[4]       # End position\n",
    "        onsen_name = onsen_row[8].split(';')[0].split('=')[1]  # Extract ONSEN name from column 8\n",
    "        \n",
    "        # Define flanking regions\n",
    "        left_flank_start = start - FLANKING_SIZE\n",
    "        left_flank_end = start\n",
    "        right_flank_start = end\n",
    "        right_flank_end = end + FLANKING_SIZE\n",
    "        \n",
    "        # Initialize a list to store methylation values for each bin (for barcode 1)\n",
    "        onsen_bins_barcode1 = [[] for _ in range((FLANKING_SIZE * 2) // BIN_SIZE)]  # Total bins = 2000 * 2 / 50 = 80\n",
    "        \n",
    "        # Process methylation data for barcode 1\n",
    "        for bin_start in range(left_flank_start, left_flank_end, BIN_SIZE):\n",
    "            bin_end = bin_start + BIN_SIZE\n",
    "            # Filter methylation data for the current bin\n",
    "            bin_data = df1[\n",
    "                (df1[0] == chr_name) &  # Match chromosome\n",
    "                (df1[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df1[2] <= bin_end)      # End position <= bin end\n",
    "            ]\n",
    "            # Append methylation values to the corresponding bin\n",
    "            bin_idx = (bin_start - left_flank_start) // BIN_SIZE\n",
    "            onsen_bins_barcode1[bin_idx].extend(bin_data[10].tolist())  # Assuming column 10 contains methylation values\n",
    "        \n",
    "        for bin_start in range(right_flank_start, right_flank_end, BIN_SIZE):\n",
    "            bin_end = bin_start + BIN_SIZE\n",
    "            # Filter methylation data for the current bin\n",
    "            bin_data = df1[\n",
    "                (df1[0] == chr_name) &  # Match chromosome\n",
    "                (df1[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df1[2] <= bin_end)      # End position <= bin end\n",
    "            ]\n",
    "            # Append methylation values to the corresponding bin\n",
    "            bin_idx = (FLANKING_SIZE // BIN_SIZE) + (bin_start - right_flank_start) // BIN_SIZE\n",
    "            onsen_bins_barcode1[bin_idx].extend(bin_data[10].tolist())  # Assuming column 10 contains methylation values\n",
    "        \n",
    "        # Add ONSEN name and binned methylation (barcode 1) to the results\n",
    "        combined_binned_methylation.append([onsen_name, onsen_bins_barcode1])\n",
    "    \n",
    "    # Iterate through ONSENs in barcode 2\n",
    "    for _, onsen_row in onsens_barcode2.iterrows():\n",
    "        # Extract ONSEN coordinates\n",
    "        chr_name = onsen_row[0]  # Chromosome\n",
    "        start = onsen_row[3]     # Start position\n",
    "        end = onsen_row[4]       # End position\n",
    "        onsen_name = onsen_row[8].split(';')[0].split('=')[1]  # Extract ONSEN name from column 8\n",
    "        \n",
    "        # Define flanking regions\n",
    "        left_flank_start = start - FLANKING_SIZE\n",
    "        left_flank_end = start\n",
    "        right_flank_start = end\n",
    "        right_flank_end = end + FLANKING_SIZE\n",
    "        \n",
    "        # Initialize a list to store methylation values for each bin (for barcode 2)\n",
    "        onsen_bins_barcode2 = [[] for _ in range((FLANKING_SIZE * 2) // BIN_SIZE)]  # Total bins = 2000 * 2 / 50 = 80\n",
    "        \n",
    "        # Process methylation data for barcode 2\n",
    "        for bin_start in range(left_flank_start, left_flank_end, BIN_SIZE):\n",
    "            bin_end = bin_start + BIN_SIZE\n",
    "            # Filter methylation data for the current bin\n",
    "            bin_data = df2[\n",
    "                (df2[0] == chr_name) &  # Match chromosome\n",
    "                (df2[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df2[2] <= bin_end)      # End position <= bin end\n",
    "            ]\n",
    "            # Append methylation values to the corresponding bin\n",
    "            bin_idx = (bin_start - left_flank_start) // BIN_SIZE\n",
    "            onsen_bins_barcode2[bin_idx].extend(bin_data[10].tolist())  # Assuming column 10 contains methylation values\n",
    "        \n",
    "        for bin_start in range(right_flank_start, right_flank_end, BIN_SIZE):\n",
    "            bin_end = bin_start + BIN_SIZE\n",
    "            # Filter methylation data for the current bin\n",
    "            bin_data = df2[\n",
    "                (df2[0] == chr_name) &  # Match chromosome\n",
    "                (df2[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df2[2] <= bin_end)      # End position <= bin end\n",
    "            ]\n",
    "            # Append methylation values to the corresponding bin\n",
    "            bin_idx = (FLANKING_SIZE // BIN_SIZE) + (bin_start - right_flank_start) // BIN_SIZE\n",
    "            onsen_bins_barcode2[bin_idx].extend(bin_data[10].tolist())  # Assuming column 10 contains methylation values\n",
    "        \n",
    "        # Add ONSEN name and binned methylation (barcode 2) to the results\n",
    "        combined_binned_methylation.append([onsen_name, onsen_bins_barcode2])\n",
    "    \n",
    "    # Combine bins from df1 and df2 for each ONSEN and calculate mean methylation\n",
    "    final_binned_methylation = []\n",
    "    for onsen_name, bins_barcode1 in combined_binned_methylation:\n",
    "        # Find corresponding bins for the same ONSEN in barcode 2\n",
    "        bins_barcode2 = next((bins for name, bins in combined_binned_methylation if name == onsen_name), None)\n",
    "        \n",
    "        if bins_barcode2:\n",
    "            # Combine methylation values bin-wise\n",
    "            combined_bins = []\n",
    "            for b1_meth, b2_meth in zip(bins_barcode1, bins_barcode2):\n",
    "                combined_values = b1_meth + b2_meth  # Combine methylation values from both barcodes\n",
    "                mean_meth = np.mean(combined_values) if combined_values else 0  # Calculate mean methylation\n",
    "                combined_bins.append(mean_meth)\n",
    "            \n",
    "            # Add ONSEN name and combined binned methylation to the results\n",
    "            final_binned_methylation.append([onsen_name] + combined_bins)\n",
    "    \n",
    "    # Convert the results to a DataFrame\n",
    "    columns = ['ONSEN'] + [f'bin_{i+1}' for i in range((FLANKING_SIZE * 2) // BIN_SIZE)]\n",
    "    results_df = pd.DataFrame(final_binned_methylation, columns=columns)\n",
    "    \n",
    "    # Store the results in the dictionary\n",
    "    results_flank_dict[sample_context] = results_df\n",
    "\n",
    "    # Save the results to CSV files\n",
    "    results_df.to_csv(f'ONSENs_flanks/ONSEN_flanking_methylation_{sample_context}.csv', index=False)\n",
    "    print(sample_context)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Binned methylation in TEIs\n",
    "## TEI body"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dict = {\n",
    "    'Col-0_CG': (meth24_cg, nte24, meth25_cg, nte25),  # CG context, barcodes 24-25\n",
    "    'Col-0_CHG': (meth24_chg, nte24, meth25_chg, nte25),  # CHG context, barcodes 24-25\n",
    "    'Col-0_CHH': (meth24_chh, nte24, meth25_chh, nte25),   # CHH context, barcodes 24-25\n",
    "\n",
    "    '12-1_CG': (meth01_cg, nte01, meth29_cg, nte29),  # CG context, barcodes 01-29\n",
    "    '12-1_CHG': (meth01_chg, nte01, meth29_chg, nte29),  # CHG context, barcodes 01-29\n",
    "    '12-1_CHH': (meth01_chh, nte01, meth29_chh, nte29),  # CHH context, barcodes 01-29\n",
    "\n",
    "    '1-1_CG': (meth19_cg, nte19, meth26_cg, nte26), \n",
    "    '1-1_CHG': (meth19_chg, nte19, meth26_chg, nte26),  \n",
    "    '1-1_CHH': (meth19_chh, nte19, meth26_chh, nte26), \n",
    "\n",
    "    '3-6_CG': (meth20_cg, nte20, meth27_cg, nte27), \n",
    "    '3-6_CHG': (meth20_chg, nte20, meth27_chg, nte27),  \n",
    "    '3-6_CHH': (meth20_chh, nte20, meth27_chh, nte27), \n",
    "\n",
    "    '7-5_CG': (meth21_cg, nte21, meth28_cg, nte28), \n",
    "    '7-5_CHG': (meth21_chg, nte21, meth28_chg, nte28),  \n",
    "    '7-5_CHH': (meth21_chh, nte21, meth28_chh, nte28), \n",
    "\n",
    "    '13-1_CG': (meth22_cg, nte22, meth30_cg, nte30), \n",
    "    '13-1_CHG': (meth22_chg, nte22, meth30_chg, nte30),  \n",
    "    '13-1_CHH': (meth22_chh, nte22, meth30_chh, nte30), \n",
    "\n",
    "    '14-2_CG': (meth23_cg, nte23), \n",
    "    '14-2_CHG': (meth23_chg, nte23),  \n",
    "    '14-2_CHH': (meth23_chh, nte23), \n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the number of bins\n",
    "NUM_BINS = 100\n",
    "\n",
    "# Initialize a dictionary to store results for each sample and context\n",
    "results_tei_dict = {}\n",
    "\n",
    "# Iterate through each sample and context\n",
    "for sample_context, data in data_dict.items():\n",
    "    sample, context = sample_context.split('_')  # Split into sample and context\n",
    "    \n",
    "    # Unpack the data for the current sample and context\n",
    "    if len(data) == 4:\n",
    "        df1_a, df2_a, df1_b, df2_b = data  # Two barcodes\n",
    "    else:\n",
    "        df1_a, df2_a = data  # One barcode\n",
    "        df1_b, df2_b = None, None  # No second barcode\n",
    "    \n",
    "    # Initialize a list to store results for this sample\n",
    "    sample_results = []\n",
    "    \n",
    "    # Iterate through each insertion in the nte dataframe\n",
    "    for _, insertion_row in df2_a.iterrows():\n",
    "        chr_name = insertion_row[0]  # Chromosome name\n",
    "        left_border = insertion_row[1]  # Left border of the insertion\n",
    "        right_border = insertion_row[2]  # Right border of the insertion\n",
    "        \n",
    "        # Calculate bin size for the current TEI\n",
    "        bin_size = (right_border - left_border) / NUM_BINS\n",
    "        \n",
    "        # Initialize a list to store combined methylation for each bin\n",
    "        combined_bin_means = []\n",
    "        \n",
    "        # Process the TEI region\n",
    "        for bin_idx in range(NUM_BINS):\n",
    "            bin_start = int(left_border + bin_idx * bin_size)  # Convert to integer\n",
    "            bin_end = int(left_border + (bin_idx + 1) * bin_size)  # Convert to integer\n",
    "            \n",
    "            # Filter df1_a for rows within the current bin\n",
    "            bin_data_a = df1_a[\n",
    "                (df1_a[0] == chr_name) &  # Match chromosome\n",
    "                (df1_a[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df1_a[2] <= bin_end)  # End position <= bin end\n",
    "            ]\n",
    "            \n",
    "            # Filter df1_b for rows within the current bin (if it exists)\n",
    "            if df1_b is not None:\n",
    "                bin_data_b = df1_b[\n",
    "                    (df1_b[0] == chr_name) &  # Match chromosome\n",
    "                    (df1_b[1] >= bin_start) &  # Start position >= bin start\n",
    "                    (df1_b[2] <= bin_end)  # End position <= bin end\n",
    "                ]\n",
    "            else:\n",
    "                bin_data_b = pd.DataFrame()  # Empty DataFrame if no second barcode\n",
    "            \n",
    "            # Combine methylation levels from both barcodes\n",
    "            if df1_b is not None:\n",
    "                # Combine methylation levels from both barcodes\n",
    "                combined_methylation = pd.concat([bin_data_a[10], bin_data_b[10]], ignore_index=True)\n",
    "            else:\n",
    "                combined_methylation = bin_data_a[10]\n",
    "            \n",
    "            # Calculate mean methylation for the bin\n",
    "            mean_methylation = combined_methylation.mean() if not combined_methylation.empty else 0\n",
    "            combined_bin_means.append(mean_methylation)\n",
    "        \n",
    "        # Append the results for this insertion\n",
    "        sample_results.append([insertion_row[3]] + combined_bin_means)  # Add TEI name as the first column\n",
    "    \n",
    "    # Convert the results to a DataFrame\n",
    "    columns = ['TEI'] + [f'bin_{i+1}' for i in range(NUM_BINS)]\n",
    "    results_df = pd.DataFrame(sample_results, columns=columns)\n",
    "    \n",
    "    # Store the results in the dictionary\n",
    "    results_tei_dict[sample_context] = results_df\n",
    "\n",
    "    # Save the results to CSV files\n",
    "    results_df.to_csv(f'TEI_meth_bins/TEI_binned_methylation_{sample_context}.csv', index=False)\n",
    "    print(sample_context)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TEI flanks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the bin size and flanking region size\n",
    "BIN_SIZE = 50\n",
    "FLANKING_SIZE = 2000\n",
    "\n",
    "# Initialize a dictionary to store results for each sample\n",
    "results_dict = {}\n",
    "\n",
    "# Iterate through each sample and context\n",
    "for sample_context, data in data_dict.items():\n",
    "    sample, context = sample_context.split('_')  # Split into sample and context\n",
    "    \n",
    "    # Unpack the data for the current sample and context\n",
    "    if len(data) == 4:\n",
    "        df1_a, df2_a, df1_b, df2_b = data  # Two barcodes\n",
    "    else:\n",
    "        df1_a, df2_a = data  # One barcode\n",
    "        df1_b, df2_b = None, None  # No second barcode\n",
    "    \n",
    "    # Initialize a list to store results for this sample\n",
    "    sample_results = []\n",
    "    \n",
    "    # Iterate through each insertion in the nte dataframe\n",
    "    for _, insertion_row in df2_a.iterrows():\n",
    "        chr_name = insertion_row[0]  # Chromosome name\n",
    "        left_border = insertion_row[1]  # Left border of the insertion\n",
    "        right_border = insertion_row[2]  # Right border of the insertion\n",
    "        \n",
    "        # Define the flanking regions\n",
    "        left_flank_start = left_border - FLANKING_SIZE\n",
    "        left_flank_end = left_border\n",
    "        right_flank_start = right_border\n",
    "        right_flank_end = right_border + FLANKING_SIZE\n",
    "        \n",
    "        # Initialize a list to store combined methylation for each bin\n",
    "        combined_bin_means = []\n",
    "        \n",
    "        # Process the left flanking region\n",
    "        for bin_start in range(left_flank_start, left_flank_end, BIN_SIZE):\n",
    "            bin_end = bin_start + BIN_SIZE\n",
    "            \n",
    "            # Filter df1_a for rows within the current bin\n",
    "            bin_data_a = df1_a[\n",
    "                (df1_a[0] == chr_name) &  # Match chromosome\n",
    "                (df1_a[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df1_a[2] <= bin_end)  # End position <= bin end\n",
    "            ]\n",
    "            \n",
    "            # Filter df1_b for rows within the current bin (if it exists)\n",
    "            if df1_b is not None:\n",
    "                bin_data_b = df1_b[\n",
    "                    (df1_b[0] == chr_name) &  # Match chromosome\n",
    "                    (df1_b[1] >= bin_start) &  # Start position >= bin start\n",
    "                    (df1_b[2] <= bin_end)  # End position <= bin end\n",
    "                ]\n",
    "            else:\n",
    "                bin_data_b = pd.DataFrame()  # Empty DataFrame if no second barcode\n",
    "\n",
    "            if df1_b is not None:\n",
    "                # Combine methylation levels from both barcodes\n",
    "                combined_methylation = pd.concat([bin_data_a[10], bin_data_b[10]], ignore_index=True)\n",
    "\n",
    "            else:\n",
    "                combined_methylation = bin_data_a[10]\n",
    "            \n",
    "            # Calculate mean methylation for the bin\n",
    "            mean_methylation = combined_methylation.mean() if not combined_methylation.empty else 0\n",
    "            combined_bin_means.append(mean_methylation)\n",
    "        \n",
    "        # Process the right flanking region\n",
    "        for bin_start in range(right_flank_start, right_flank_end, BIN_SIZE):\n",
    "            bin_end = bin_start + BIN_SIZE\n",
    "            \n",
    "            # Filter df1_a for rows within the current bin\n",
    "            bin_data_a = df1_a[\n",
    "                (df1_a[0] == chr_name) &  # Match chromosome\n",
    "                (df1_a[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df1_a[2] <= bin_end)  # End position <= bin end\n",
    "            ]\n",
    "            \n",
    "            # Filter df1_b for rows within the current bin (if it exists)\n",
    "            if df1_b is not None:\n",
    "                bin_data_b = df1_b[\n",
    "                    (df1_b[0] == chr_name) &  # Match chromosome\n",
    "                    (df1_b[1] >= bin_start) &  # Start position >= bin start\n",
    "                    (df1_b[2] <= bin_end)  # End position <= bin end\n",
    "                ]\n",
    "            else:\n",
    "                bin_data_b = pd.DataFrame()  # Empty DataFrame if no second barcode\n",
    "            \n",
    "            if df1_b is not None:\n",
    "                # Combine methylation levels from both barcodes\n",
    "                combined_methylation = pd.concat([bin_data_a[10], bin_data_b[10]], ignore_index=True)\n",
    "            else:\n",
    "                combined_methylation = bin_data_a[10]\n",
    "            \n",
    "            # Calculate mean methylation for the bin\n",
    "            mean_methylation = combined_methylation.mean() if not combined_methylation.empty else 0\n",
    "            combined_bin_means.append(mean_methylation)\n",
    "        \n",
    "        # Append the results for this insertion\n",
    "        sample_results.append(combined_bin_means)\n",
    "    \n",
    "    # Convert the results to a DataFrame\n",
    "    columns = [f'bin_{i+1}' for i in range(80)]  # 80 bins\n",
    "    results_df = pd.DataFrame(sample_results, columns=columns)\n",
    "    results_df['insertion'] = df2_a[3]\n",
    "    # results_df.to_csv(f'TEI_flanks/{sample_context}_flanking_methylation_2kb.csv', index=False)\n",
    "\n",
    "    # Store the results in the dictionary\n",
    "    results_dict[sample_context] = results_df\n",
    "    print(sample_context)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TEI flanks in Col-0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# unshifted methylation (tair10)\n",
    "meth24_cg = pd.read_csv('../ont_methylation/sorted_SQK-NBD114-96_barcode24.meth.tair10.CG.bed', sep='\\t', header=None)\n",
    "meth24_chg = pd.read_csv('../ont_methylation/sorted_SQK-NBD114-96_barcode24.meth.tair10.CHG.bed', sep='\\t', header=None)\n",
    "meth24_chh = pd.read_csv('../ont_methylation/sorted_SQK-NBD114-96_barcode24.meth.tair10.CHH.bed', sep='\\t', header=None)\n",
    "\n",
    "meth25_cg = pd.read_csv('../ont_methylation/sorted_SQK-NBD114-96_barcode25.meth.tair10.CG.bed', sep='\\t', header=None)\n",
    "meth25_chg = pd.read_csv('../ont_methylation/sorted_SQK-NBD114-96_barcode25.meth.tair10.CHG.bed', sep='\\t', header=None)\n",
    "meth25_chh = pd.read_csv('../ont_methylation/sorted_SQK-NBD114-96_barcode25.meth.tair10.CHH.bed', sep='\\t', header=None)\n",
    "\n",
    "df_list = [meth24_cg, meth24_chg, meth24_chh, \n",
    "           meth25_cg, meth25_chg, meth25_chh]\n",
    "name_list = ['meth24_cg', 'meth24_chg', 'meth24_chh', \n",
    "           'meth25_cg', 'meth25_chg', 'meth25_chh']\n",
    "filt_df = {}\n",
    "\n",
    "for df, name in zip(df_list, name_list):\n",
    "    filt_df[name] = df[(df[9] <= df[9].quantile(0.95)) & (df[9] >= df[9].quantile(0.05))]\n",
    "\n",
    "meth24_cg = filt_df['meth24_cg']\n",
    "meth24_chg = filt_df['meth24_chg']\n",
    "meth24_chh = filt_df['meth24_chh']\n",
    "meth25_cg = filt_df['meth25_cg']\n",
    "meth25_chg = filt_df['meth25_chg']\n",
    "meth25_chh = filt_df['meth25_chh']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# unshifted novel insertions positions\n",
    "tei_df = pd.read_csv('TEI_S1_pseudoref_ins.tsv', sep='\\t')\n",
    "tei_df = tei_df[tei_df['VI']==True]\n",
    "tei_df['Corrected pos'].fillna(tei_df['TEIcoord'], inplace=True)\n",
    "tei_df['TEIcoord_chr'] = [n[:11] for n in tei_df['Corrected pos'].values]\n",
    "tei_df['TEIcoord_start'] = [n.split(':')[1].split('..')[0] for n in tei_df['Corrected pos'].values]\n",
    "tei_df['TEIcoord_end'] = [n.split(':')[1].split('..')[1] for n in tei_df['Corrected pos'].values]\n",
    "\n",
    "tei_df['TEIcoord_start'] = tei_df['TEIcoord_start'].astype(int)\n",
    "tei_df['TEIcoord_end'] = tei_df['TEIcoord_end'].astype(int)\n",
    "\n",
    "tei_df = tei_df.sort_values(['sample', 'TEIcoord_chr', 'TEIcoord_start'], ascending=[True, True, True])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "nte01 = tei_df[tei_df['sample']=='SQK-NBD114-96_barcode01'][['TEIcoord_chr', 'TEIcoord_start', 'TEIcoord_end', 'Corrected pos']]\n",
    "nte01.columns = [0, 1, 2, 3]\n",
    "nte01.reset_index(inplace=True, drop=True)\n",
    "nte19 = tei_df[tei_df['sample']=='SQK-NBD114-96_barcode19'][['TEIcoord_chr', 'TEIcoord_start', 'TEIcoord_end', 'Corrected pos']]\n",
    "nte19.columns = [0, 1, 2, 3]\n",
    "nte19.reset_index(inplace=True)\n",
    "nte20 = tei_df[tei_df['sample']=='SQK-NBD114-96_barcode20'][['TEIcoord_chr', 'TEIcoord_start', 'TEIcoord_end', 'Corrected pos']]\n",
    "nte20.columns = [0, 1, 2, 3]\n",
    "nte20.reset_index(inplace=True)\n",
    "nte21 = tei_df[tei_df['sample']=='SQK-NBD114-96_barcode21'][['TEIcoord_chr', 'TEIcoord_start', 'TEIcoord_end', 'Corrected pos']]\n",
    "nte21.columns = [0, 1, 2, 3]\n",
    "nte21.reset_index(inplace=True)\n",
    "nte22 = tei_df[tei_df['sample']=='SQK-NBD114-96_barcode22'][['TEIcoord_chr', 'TEIcoord_start', 'TEIcoord_end', 'Corrected pos']]\n",
    "nte22.columns = [0, 1, 2, 3]\n",
    "nte22.reset_index(inplace=True)\n",
    "nte23 = tei_df[tei_df['sample']=='SQK-NBD114-96_barcode23'][['TEIcoord_chr', 'TEIcoord_start', 'TEIcoord_end', 'Corrected pos']]\n",
    "nte23.columns = [0, 1, 2, 3]\n",
    "nte23.reset_index(inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dict = {\n",
    "    'Col-0_CG': (meth24_cg, meth25_cg),  # CG context, barcodes 24-25\n",
    "    'Col-0_CHG': (meth24_chg, meth25_chg),  # CHG context, barcodes 24-25\n",
    "    'Col-0_CHH': (meth24_chh, meth25_chh),   # CHH context, barcodes 24-25\n",
    "}\n",
    "\n",
    "ins_dict = {\n",
    "            # '12-1': nte01,\n",
    "            '1-1': nte19, \n",
    "            '3-6': nte20, \n",
    "            '7-5': nte21, \n",
    "            '13-1': nte22, \n",
    "            '14-2': nte23\n",
    "            }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1-1_Col-0_CG\n",
      "1-1_Col-0_CHG\n",
      "1-1_Col-0_CHH\n",
      "3-6_Col-0_CG\n",
      "3-6_Col-0_CHG\n",
      "3-6_Col-0_CHH\n",
      "7-5_Col-0_CG\n",
      "7-5_Col-0_CHG\n",
      "7-5_Col-0_CHH\n",
      "13-1_Col-0_CG\n",
      "13-1_Col-0_CHG\n",
      "13-1_Col-0_CHH\n",
      "14-2_Col-0_CG\n",
      "14-2_Col-0_CHG\n",
      "14-2_Col-0_CHH\n"
     ]
    }
   ],
   "source": [
    "# Define the bin size and flanking region size\n",
    "BIN_SIZE = 50\n",
    "FLANKING_SIZE = 2000\n",
    "NUM_BINS = int(2 * FLANKING_SIZE / BIN_SIZE) # 40 bins per flank (2*2000/50 = 80)\n",
    "\n",
    "# Initialize a dictionary to store results for each insertion and context\n",
    "results_dict = {}\n",
    "\n",
    "# Iterate through each insertion dataframe\n",
    "for ins_id, ins_df in ins_dict.items():\n",
    "    # Iterate through each context (Col-0_CG, Col-0_CHG, Col-0_CHH)\n",
    "    for context, (df1_a, df1_b) in data_dict.items():\n",
    "        # Pre-compute numpy arrays for faster access\n",
    "        chr_a = df1_a[0].values\n",
    "        start_a = df1_a[1].values\n",
    "        end_a = df1_a[2].values\n",
    "        meth_a = df1_a[10].values\n",
    "        \n",
    "        chr_b = df1_b[0].values\n",
    "        start_b = df1_b[1].values\n",
    "        end_b = df1_b[2].values\n",
    "        meth_b = df1_b[10].values\n",
    "        \n",
    "        # Initialize array to store all insertion results\n",
    "        all_insertion_results = np.zeros((len(ins_df), NUM_BINS))\n",
    "        \n",
    "        # Iterate through each insertion in the insertion dataframe\n",
    "        for ins_idx, insertion_row in enumerate(ins_df.itertuples()):\n",
    "            # Extract chromosome and borders from the insertion position (column 3)\n",
    "            insertion_pos = insertion_row[-1]\n",
    "            chr_name, borders = insertion_pos.split(':')\n",
    "            left_border, right_border = map(int, borders.split('..'))\n",
    "            \n",
    "            # Define the flanking regions\n",
    "            left_flank_start = left_border - FLANKING_SIZE\n",
    "            right_flank_end = right_border + FLANKING_SIZE\n",
    "            \n",
    "            # Create all bin ranges at once\n",
    "            left_bins = np.arange(left_flank_start, left_border, BIN_SIZE)\n",
    "            right_bins = np.arange(right_border, right_flank_end, BIN_SIZE)\n",
    "            all_bins = np.concatenate([left_bins, right_bins])\n",
    "            \n",
    "            # Initialize array to store bin means\n",
    "            bin_means = np.zeros(NUM_BINS)\n",
    "            \n",
    "            for bin_idx, bin_start in enumerate(all_bins):\n",
    "                bin_end = bin_start + BIN_SIZE\n",
    "                \n",
    "                # Process df1_a - using numpy boolean indexing\n",
    "                mask_a = (chr_a == chr_name) & (start_a >= bin_start) & (end_a <= bin_end)\n",
    "                meth_values_a = meth_a[mask_a]\n",
    "                \n",
    "                # Process df1_b - using numpy boolean indexing\n",
    "                mask_b = (chr_b == chr_name) & (start_b >= bin_start) & (end_b <= bin_end)\n",
    "                meth_values_b = meth_b[mask_b]\n",
    "                \n",
    "                # Combine and calculate mean\n",
    "                combined = np.concatenate([meth_values_a, meth_values_b])\n",
    "                bin_means[bin_idx] = np.mean(combined) if combined.size > 0 else 0\n",
    "            \n",
    "            # Store results for this insertion\n",
    "            all_insertion_results[ins_idx] = bin_means\n",
    "        \n",
    "        # Convert to DataFrame\n",
    "        columns = [f'bin_{i+1}' for i in range(NUM_BINS)]\n",
    "        results_df = pd.DataFrame(all_insertion_results, columns=columns)\n",
    "        results_df['insertion'] = ins_df[3].values\n",
    "        \n",
    "        # Save results\n",
    "        results_df.to_csv(f'TEI_flanks/{ins_id}_{context}_flanking_methylation.csv', index=False)\n",
    "        results_dict[f'{ins_id}_{context}'] = results_df\n",
    "        print(f'{ins_id}_{context}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Binned methylation in EVD \n",
    "## TE body"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "evd24 = pd.read_csv('evd_positions/EVD_barcode24.gff', sep='\\t', header=None)\n",
    "evd01 = pd.read_csv('evd_positions/EVD_barcode01.gff', sep='\\t', header=None)\n",
    "evd25 = pd.read_csv('evd_positions/EVD_barcode25.gff', sep='\\t', header=None)\n",
    "evd29 = pd.read_csv('evd_positions/EVD_barcode29.gff', sep='\\t', header=None)\n",
    "evd19 = pd.read_csv('evd_positions/EVD_barcode19.gff', sep='\\t', header=None)\n",
    "evd20 = pd.read_csv('evd_positions/EVD_barcode20.gff', sep='\\t', header=None)\n",
    "evd21 = pd.read_csv('evd_positions/EVD_barcode21.gff', sep='\\t', header=None)\n",
    "evd22 = pd.read_csv('evd_positions/EVD_barcode22.gff', sep='\\t', header=None)\n",
    "evd23 = pd.read_csv('evd_positions/EVD_barcode23.gff', sep='\\t', header=None)\n",
    "evd26 = pd.read_csv('evd_positions/EVD_barcode26.gff', sep='\\t', header=None)\n",
    "evd27 = pd.read_csv('evd_positions/EVD_barcode27.gff', sep='\\t', header=None)\n",
    "evd28 = pd.read_csv('evd_positions/EVD_barcode28.gff', sep='\\t', header=None)\n",
    "evd30 = pd.read_csv('evd_positions/EVD_barcode30.gff', sep='\\t', header=None)\n",
    "evd31 = pd.read_csv('evd_positions/EVD_barcode31.gff', sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dict = {\n",
    "    'Col-0_CG': (meth24_cg, evd24, meth25_cg, evd25),  # CG context, barcodes 24-25\n",
    "    'Col-0_CHG': (meth24_chg, evd24, meth25_chg, evd25),  # CHG context, barcodes 24-25\n",
    "    'Col-0_CHH': (meth24_chh, evd24, meth25_chh, evd25),   # CHH context, barcodes 24-25\n",
    "\n",
    "    '12-1_CG': (meth01_cg, evd01, meth29_cg, evd29),  # CG context, barcodes 01-29\n",
    "    '12-1_CHG': (meth01_chg, evd01, meth29_chg, evd29),  # CHG context, barcodes 01-29\n",
    "    '12-1_CHH': (meth01_chh, evd01, meth29_chh, evd29),  # CHH context, barcodes 01-29\n",
    "\n",
    "    '1-1_CG': (meth19_cg, evd19, meth26_cg, evd26), \n",
    "    '1-1_CHG': (meth19_chg, evd19, meth26_chg, evd26),  \n",
    "    '1-1_CHH': (meth19_chh, evd19, meth26_chh, evd26), \n",
    "\n",
    "    '3-6_CG': (meth20_cg, evd20, meth27_cg, evd27), \n",
    "    '3-6_CHG': (meth20_chg, evd20, meth27_chg, evd27),  \n",
    "    '3-6_CHH': (meth20_chh, evd20, meth27_chh, evd27), \n",
    "\n",
    "    '7-5_CG': (meth21_cg, evd21, meth28_cg, evd28), \n",
    "    '7-5_CHG': (meth21_chg, evd21, meth28_chg, evd28),  \n",
    "    '7-5_CHH': (meth21_chh, evd21, meth28_chh, evd28), \n",
    "\n",
    "    '13-1_CG': (meth22_cg, evd22, meth30_cg, evd30), \n",
    "    '13-1_CHG': (meth22_chg, evd22, meth30_chg, evd30),  \n",
    "    '13-1_CHH': (meth22_chh, evd22, meth30_chh, evd30), \n",
    "\n",
    "    '14-2_CG': (meth23_cg, evd23, meth31_cg, evd31), \n",
    "    '14-2_CHG': (meth23_chg, evd23, meth31_chg, evd31),  \n",
    "    '14-2_CHH': (meth23_chh, evd23, meth31_chh, evd31), \n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Col-0_CG\n",
      "Col-0_CHG\n",
      "Col-0_CHH\n",
      "12-1_CG\n",
      "12-1_CHG\n",
      "12-1_CHH\n",
      "1-1_CG\n",
      "1-1_CHG\n",
      "1-1_CHH\n",
      "3-6_CG\n",
      "3-6_CHG\n",
      "3-6_CHH\n",
      "7-5_CG\n",
      "7-5_CHG\n",
      "7-5_CHH\n",
      "13-1_CG\n",
      "13-1_CHG\n",
      "13-1_CHH\n",
      "14-2_CG\n",
      "14-2_CHG\n",
      "14-2_CHH\n"
     ]
    }
   ],
   "source": [
    "# Define the number of bins\n",
    "NUM_BINS = 100\n",
    "\n",
    "# Initialize a dictionary to store results for each sample and context\n",
    "results_dict = {}\n",
    "\n",
    "# Iterate through each sample and context\n",
    "for sample_context, (df1, evd1, df2, evd2) in data_dict.items():\n",
    "    # Initialize a list to store combined binned methylation for all EVDs\n",
    "    combined_binned_methylation = []\n",
    "    \n",
    "    # Get unique EVDs for each barcode\n",
    "    evds_barcode1 = evd1.drop_duplicates().reset_index(drop=True)\n",
    "    evds_barcode2 = evd2.drop_duplicates().reset_index(drop=True)\n",
    "    \n",
    "    # Iterate through EVDs in barcode 1\n",
    "    for _, evd_row in evds_barcode1.iterrows():\n",
    "        # Extract EVD coordinates\n",
    "        chr_name = evd_row[0]  # Chromosome\n",
    "        start = evd_row[3]     # Start position\n",
    "        end = evd_row[4]       # End position\n",
    "        evd_name = evd_row[8].split(';')[0].split('=')[1]  # Extract EVD name from column 8\n",
    "        \n",
    "        # Initialize lists to store combined methylation values and counts for each bin\n",
    "        combined_bins_meth = [[] for _ in range(NUM_BINS)]\n",
    "        combined_bins_count = [0] * NUM_BINS\n",
    "        \n",
    "        # Calculate bin size for the current EVD\n",
    "        bin_size = (end - start) / NUM_BINS\n",
    "        \n",
    "        # Process methylation data for barcode 1\n",
    "        evd_meth_barcode1 = df1[\n",
    "            (df1[0] == chr_name) &  # Match chromosome\n",
    "            (df1[1] >= start) &     # Start position >= EVD start\n",
    "            (df1[2] <= end)         # End position <= EVD end\n",
    "        ]\n",
    "        for _, row in evd_meth_barcode1.iterrows():\n",
    "            pos = row[1]  # Methylation position\n",
    "            bin_idx = min(int((pos - start) // bin_size), NUM_BINS - 1)  # Ensure bin index is within range\n",
    "            combined_bins_meth[bin_idx].append(row[10])  # Assuming column 10 contains methylation values\n",
    "            combined_bins_count[bin_idx] += 1  # Count cytosines\n",
    "        \n",
    "        # Process methylation data for barcode 2\n",
    "        evd_meth_barcode2 = df2[\n",
    "            (df2[0] == chr_name) &  # Match chromosome\n",
    "            (df2[1] >= start) &     # Start position >= EVD start\n",
    "            (df2[2] <= end)         # End position <= EVD end\n",
    "        ]\n",
    "        for _, row in evd_meth_barcode2.iterrows():\n",
    "            pos = row[1]  # Methylation position\n",
    "            bin_idx = min(int((pos - start) // bin_size), NUM_BINS - 1)  # Ensure bin index is within range\n",
    "            combined_bins_meth[bin_idx].append(row[10])  # Assuming column 10 contains methylation values\n",
    "            combined_bins_count[bin_idx] += 1  # Count cytosines\n",
    "        \n",
    "        # Calculate mean methylation for each bin\n",
    "        mean_binned_meth = []\n",
    "        for meth_values, count in zip(combined_bins_meth, combined_bins_count):\n",
    "            if count > 0:\n",
    "                mean_meth = np.mean(meth_values)  # Calculate mean methylation\n",
    "            else:\n",
    "                mean_meth = 0  # If no cytosines, set methylation to 0\n",
    "            mean_binned_meth.append(mean_meth)\n",
    "        \n",
    "        # Add EVD name and mean binned methylation to the results\n",
    "        combined_binned_methylation.append([evd_name] + mean_binned_meth)\n",
    "    \n",
    "    # Convert the results to a DataFrame\n",
    "    columns = ['EVD'] + [f'bin_{i+1}' for i in range(NUM_BINS)]\n",
    "    results_df = pd.DataFrame(combined_binned_methylation, columns=columns)\n",
    "    \n",
    "    # Store the results in the dictionary\n",
    "    results_dict[sample_context] = results_df\n",
    "\n",
    "    # Save the results to CSV files\n",
    "    results_df.to_csv(f'EVD_meth_bins/EVD_binned_methylation_{sample_context}.csv', index=False)\n",
    "    print(sample_context)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TE flanks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Col-0_CG\n",
      "Col-0_CHG\n",
      "Col-0_CHH\n",
      "12-1_CG\n",
      "12-1_CHG\n",
      "12-1_CHH\n",
      "1-1_CG\n",
      "1-1_CHG\n",
      "1-1_CHH\n",
      "3-6_CG\n",
      "3-6_CHG\n",
      "3-6_CHH\n",
      "7-5_CG\n",
      "7-5_CHG\n",
      "7-5_CHH\n",
      "13-1_CG\n",
      "13-1_CHG\n",
      "13-1_CHH\n",
      "14-2_CG\n",
      "14-2_CHG\n",
      "14-2_CHH\n"
     ]
    }
   ],
   "source": [
    "# Define the bin size and flanking size\n",
    "BIN_SIZE = 50  # Fixed bin size\n",
    "FLANKING_SIZE = 2000  # 2000 bp flanking regions\n",
    "\n",
    "# Initialize a dictionary to store results for each sample and context\n",
    "results_flank_dict = {}\n",
    "\n",
    "# Iterate through each sample and context\n",
    "for sample_context, (df1, evd1, df2, evd2) in data_dict.items():\n",
    "    # Initialize a list to store combined binned methylation for all EVDs\n",
    "    combined_binned_methylation = []\n",
    "    \n",
    "    # Get unique EVDs for each barcode\n",
    "    evds_barcode1 = evd1.drop_duplicates().reset_index(drop=True)\n",
    "    evds_barcode2 = evd2.drop_duplicates().reset_index(drop=True)\n",
    "    \n",
    "    # Iterate through EVDs in barcode 1\n",
    "    for _, evd_row in evds_barcode1.iterrows():\n",
    "        # Extract EVD coordinates\n",
    "        chr_name = evd_row[0]  # Chromosome\n",
    "        start = evd_row[3]     # Start position\n",
    "        end = evd_row[4]       # End position\n",
    "        evd_name = evd_row[8].split(';')[0].split('=')[1]  # Extract EVD name from column 8\n",
    "        \n",
    "        # Define flanking regions\n",
    "        left_flank_start = start - FLANKING_SIZE\n",
    "        left_flank_end = start\n",
    "        right_flank_start = end\n",
    "        right_flank_end = end + FLANKING_SIZE\n",
    "        \n",
    "        # Initialize a list to store methylation values for each bin (for barcode 1)\n",
    "        evd_bins_barcode1 = [[] for _ in range((FLANKING_SIZE * 2) // BIN_SIZE)]  # Total bins = 2000 * 2 / 50 = 80\n",
    "        \n",
    "        # Process methylation data for barcode 1\n",
    "        for bin_start in range(left_flank_start, left_flank_end, BIN_SIZE):\n",
    "            bin_end = bin_start + BIN_SIZE\n",
    "            # Filter methylation data for the current bin\n",
    "            bin_data = df1[\n",
    "                (df1[0] == chr_name) &  # Match chromosome\n",
    "                (df1[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df1[2] <= bin_end)      # End position <= bin end\n",
    "            ]\n",
    "            # Append methylation values to the corresponding bin\n",
    "            bin_idx = (bin_start - left_flank_start) // BIN_SIZE\n",
    "            evd_bins_barcode1[bin_idx].extend(bin_data[10].tolist())  # Assuming column 10 contains methylation values\n",
    "        \n",
    "        for bin_start in range(right_flank_start, right_flank_end, BIN_SIZE):\n",
    "            bin_end = bin_start + BIN_SIZE\n",
    "            # Filter methylation data for the current bin\n",
    "            bin_data = df1[\n",
    "                (df1[0] == chr_name) &  # Match chromosome\n",
    "                (df1[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df1[2] <= bin_end)      # End position <= bin end\n",
    "            ]\n",
    "            # Append methylation values to the corresponding bin\n",
    "            bin_idx = (FLANKING_SIZE // BIN_SIZE) + (bin_start - right_flank_start) // BIN_SIZE\n",
    "            evd_bins_barcode1[bin_idx].extend(bin_data[10].tolist())  # Assuming column 10 contains methylation values\n",
    "        \n",
    "        # Add EVD name and binned methylation (barcode 1) to the results\n",
    "        combined_binned_methylation.append([evd_name, evd_bins_barcode1])\n",
    "    \n",
    "    # Iterate through EVDs in barcode 2\n",
    "    for _, evd_row in evds_barcode2.iterrows():\n",
    "        # Extract EVD coordinates\n",
    "        chr_name = evd_row[0]  # Chromosome\n",
    "        start = evd_row[3]     # Start position\n",
    "        end = evd_row[4]       # End position\n",
    "        evd_name = evd_row[8].split(';')[0].split('=')[1]  # Extract EVD name from column 8\n",
    "        \n",
    "        # Define flanking regions\n",
    "        left_flank_start = start - FLANKING_SIZE\n",
    "        left_flank_end = start\n",
    "        right_flank_start = end\n",
    "        right_flank_end = end + FLANKING_SIZE\n",
    "        \n",
    "        # Initialize a list to store methylation values for each bin (for barcode 2)\n",
    "        evd_bins_barcode2 = [[] for _ in range((FLANKING_SIZE * 2) // BIN_SIZE)]  # Total bins = 2000 * 2 / 50 = 80\n",
    "        \n",
    "        # Process methylation data for barcode 2\n",
    "        for bin_start in range(left_flank_start, left_flank_end, BIN_SIZE):\n",
    "            bin_end = bin_start + BIN_SIZE\n",
    "            # Filter methylation data for the current bin\n",
    "            bin_data = df2[\n",
    "                (df2[0] == chr_name) &  # Match chromosome\n",
    "                (df2[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df2[2] <= bin_end)      # End position <= bin end\n",
    "            ]\n",
    "            # Append methylation values to the corresponding bin\n",
    "            bin_idx = (bin_start - left_flank_start) // BIN_SIZE\n",
    "            evd_bins_barcode2[bin_idx].extend(bin_data[10].tolist())  # Assuming column 10 contains methylation values\n",
    "        \n",
    "        for bin_start in range(right_flank_start, right_flank_end, BIN_SIZE):\n",
    "            bin_end = bin_start + BIN_SIZE\n",
    "            # Filter methylation data for the current bin\n",
    "            bin_data = df2[\n",
    "                (df2[0] == chr_name) &  # Match chromosome\n",
    "                (df2[1] >= bin_start) &  # Start position >= bin start\n",
    "                (df2[2] <= bin_end)      # End position <= bin end\n",
    "            ]\n",
    "            # Append methylation values to the corresponding bin\n",
    "            bin_idx = (FLANKING_SIZE // BIN_SIZE) + (bin_start - right_flank_start) // BIN_SIZE\n",
    "            evd_bins_barcode2[bin_idx].extend(bin_data[10].tolist())  # Assuming column 10 contains methylation values\n",
    "        \n",
    "        # Add EVD name and binned methylation (barcode 2) to the results\n",
    "        combined_binned_methylation.append([evd_name, evd_bins_barcode2])\n",
    "    \n",
    "    # Combine bins from df1 and df2 for each EVD and calculate mean methylation\n",
    "    final_binned_methylation = []\n",
    "    for evd_name, bins_barcode1 in combined_binned_methylation:\n",
    "        # Find corresponding bins for the same EVD in barcode 2\n",
    "        bins_barcode2 = next((bins for name, bins in combined_binned_methylation if name == evd_name), None)\n",
    "        \n",
    "        if bins_barcode2:\n",
    "            # Combine methylation values bin-wise\n",
    "            combined_bins = []\n",
    "            for b1_meth, b2_meth in zip(bins_barcode1, bins_barcode2):\n",
    "                combined_values = b1_meth + b2_meth  # Combine methylation values from both barcodes\n",
    "                mean_meth = np.mean(combined_values) if combined_values else 0  # Calculate mean methylation\n",
    "                combined_bins.append(mean_meth)\n",
    "            \n",
    "            # Add EVD name and combined binned methylation to the results\n",
    "            final_binned_methylation.append([evd_name] + combined_bins)\n",
    "    \n",
    "    # Convert the results to a DataFrame\n",
    "    columns = ['EVD'] + [f'bin_{i+1}' for i in range((FLANKING_SIZE * 2) // BIN_SIZE)]\n",
    "    results_df = pd.DataFrame(final_binned_methylation, columns=columns)\n",
    "    \n",
    "    # Store the results in the dictionary\n",
    "    results_flank_dict[sample_context] = results_df\n",
    "\n",
    "    # Save the results to CSV files\n",
    "    results_df.to_csv(f'EVDs_flanks/EVD_flanking_methylation_{sample_context}.csv', index=False)\n",
    "    print(sample_context)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
